File: pygmin/potentials/fortran/maxneib_lj.f90
function fermi(x, mu, T) result(ferm)
double precision, intent(in) :: x, mu, T
double precision ferm
ferm = 1.d0 / (exp(-(x - mu) / T) + 1.d0)
end function

subroutine maxneib_ljenergy( coords, natoms, e, eps, sig, periodic, boxl, &
   rneib, rneib_crossover, max_neibs, neib_crossover, epsneibs)
implicit none
integer, intent(in) :: natoms
double precision, intent(in) :: coords(3*natoms), sig, eps, boxl
double precision, intent(out) :: e
double precision dr(3), sig6, sig12, r2, ir2, ir6, ir12, iboxl
logical, intent(in) :: periodic
integer j1, j2
double precision, intent(in) :: rneib, rneib_crossover, max_neibs
double precision, intent(in) :: neib_crossover, epsneibs
double precision nneibs(natoms)
double precision eneibs, eneibs1, areneibs
double precision fermi

nneibs(:) = 0.d0

!rneib = 1.4
!rneib_crossover = .08
!max_neibs = 5.
!neib_crossover = .4
!epsneibs = 1.


if (periodic) iboxl = 1.d0 / boxl

sig6 = sig**6
sig12 = sig6*sig6

e = 0.d0
do j1 = 1,natoms
   do j2 = 1,j1-1
      dr(:) = coords(3*(j1-1)+1 : 3*(j1-1) + 3) - coords(3*(j2-1)+1 : 3*(j2-1) + 3)
      if (periodic)  dr(:) = dr(:) - nint( dr(:) * iboxl ) * boxl
      r2 = sum( dr(:)**2 )
      ir2 = 1.d0/r2
      ir6 = ir2**3
      ir12 = ir6**2
      e = e - 4.d0 * eps * (sig6*ir6 - sig12*ir12)
      areneibs = 1.d0 - fermi(sqrt(r2), rneib, rneib_crossover)
      nneibs(j1) = nneibs(j1)  + areneibs
      nneibs(j2) = nneibs(j2)  + areneibs
   enddo
enddo

!calculate energy from neighbors
eneibs = 0.d0
do j1=1,natoms
   eneibs1 = fermi(nneibs(j1), max_neibs, neib_crossover)
   eneibs = eneibs + eneibs1
enddo
eneibs = eneibs * epsneibs

e = e + eneibs

end subroutine maxneib_ljenergy

File: pygmin/potentials/maxneib_lj.py
"""
lj potential with the number of near neighbors restricted.
"""
import numpy as np

from pygmin.potentials import BasePotential
from pygmin.systems import LJCluster
import fortran.maxneib_lj as fortranpot



class MaxNeibsLJ(BasePotential):
    """
    atoms interact with lj potential, with an additional energy penalty for too many neighbors
    
    use the fermi function to get a continuous measure of whether two atoms are neighbors.  this gives
    a continuous measure of the number of neighbors of an atoms.
    
    use the fermi function again to calculate the energy penalty if an atom has too many neighbors.
    
    Parameters
    ----------
    eps : float
        lj eps
    sig : float
        lj sig
    boxl : float
        box length if periodic
    max_neibs : float 
        maximum number of allowed neighbors
    neib_crossover : float
        width of neighbor energy crossover function
    rneib : float
        distance to consider two atoms neighbors
    rneib_crossover : float
        width of rneib crossover region
    epsneibs : float
        energy scale of the neighbor penalty function
    
    
    """
    def __init__(self, eps=1.0, sig=1.0, boxl=None,
                 max_neibs=5.,
                 neib_crossover=.4,
                 rneib=1.4,
                 rneib_crossover=0.08,
                 epsneibs=5.,
                 ):
        """ simple lennard jones potential"""
        self.sig = sig
        self.eps = eps
        
        if boxl is None:
            self.boxl = 1e100
            self.periodic = False
        else:
            self.boxl = boxl
            self.periodic = True
        
        self.rneib = rneib
        self.rneib_crossover = rneib_crossover
        self.max_neibs = max_neibs
        self.neib_crossover = neib_crossover
        self.epsneibs = epsneibs

    def getEnergy(self, coords):
        E = fortranpot.maxneib_ljenergy(
                coords, self.eps, self.sig, self.periodic, self.boxl, 
                self.rneib, self.rneib_crossover, self.max_neibs, self.neib_crossover, 
                self.epsneibs)
        return E
    def getEnergyGradient(self, coords):
        E, grad = fortranpot.maxneib_ljenergy_gradient(
                coords, self.eps, self.sig, self.periodic, self.boxl, 
                self.rneib, self.rneib_crossover, self.max_neibs, self.neib_crossover, 
                self.epsneibs)
        return E, grad


class MaxNeibsLJSystem(LJCluster):
    def __init__(self, natoms, **potkwargs):
        super(MaxNeibsLJSystem, self).__init__(natoms)
        self.potkwargs = potkwargs
        self.params.gui.basinhopping_nsteps = 1000
    def __call__(self):
        return self
    
    def get_potential(self):
        return MaxNeibsLJ(**self.potkwargs)


def run_gui(system):
    import pygmin.gui.run as gr
    gr.run_gui(system)


if __name__ == "__main__":
    natoms = 20
    system = MaxNeibsLJSystem(natoms, max_neibs=3, rneib=1.7)
    
    coords = system.get_random_configuration()
    pot = system.get_potential()
    E = pot.getEnergy(coords)
    print "energy", E
    pot.test_potential(coords)
    if True:
        coords = system.get_random_minimized_configuration()[0]
        pot.test_potential(coords)
    
    run_gui(system)
